### ISSUE TITLE
Deploy curl-proxy/wget-proxy scripts so users can add GPG distro keys linked to added external repositories

### ISSUE THREAD

**Issue by tlaurion** [QubesOS/qubes-issues#7386]
EDIT: PoC [here](https://forum.qubes-os.org/t/curl-proxy-wget-proxy-scripts-in-templates-so-users-can-add-gpg-distro-keys-linked-to-added-external-repositories/10935/37?u=insurgo)

This implements wget and curl wrappers, blocking when usage happens in TemplateVM without internet access (considering that TemplateVM is considered secure). The code presented there could be modified as the next use case (TemplateVM with netvm assigned) to only pause for 3 seconds warning the user of the original wget/curl command going to happen next and telling the user to CTRL-C now if that command was not intended.

That sleeping of 3 seconds was implemented so that unattended scripts would still work out of the box, without the user having to copy one line at a time, if the wrapper scripts were to ask the user for input prior of continuing, which would have broke caller scripts (and stopping bash scripts and other to automatize software installation scripts).

I encourage people reading this to comment directly in the forum post for discussion, or here for implementation suggestion to make it more acceptable. Those threads were immense. This is the get home message/PoC for users exploration.

### The problem you're addressing (if any)

Customers/users are struggling everyday with installing software in templates that are not part of standard repositories. This won't fix the bigger problems of personas or having different classes of softwares being installed by default without manual intervention, but at least, this permits users to download gpg signing keys as a near drop in replacement to installation instructions given by upstream guides, permitting users to follow generally the installation instructions that exist, without playing around too much to apply those instructions. 

Otherwise, users  punch holes in their configuration, assigning a network vm to their templates, more or less diligently denying internet access but punching exception for it to work for the next 5 minutes only, which opens a big door to downloading other crap for users trying to launch gui-installers and possibly do important damage in their templates.

Or requiring from those users to download gpg keys in disposable qubes and having to move the gpg public key back into the template, which implies of them of being able to decode/understand installation instructions (most of them can't and that complexifies onboarding).

The solution is not to deploy snap. Users need to be able to economize disk space without having applications deployed directly from snap repositories and duplicating the network bandwidth required for all their qubes where snap packages are deployed. Users also need to know that they need to update their software, prior of launching them. Snap is confusing with no warning given that a new version is downloaded and installed but not currently running.

The solution is not to have software providers create specific Qubes installation instructions either.

### The solution you'd like
The real solution is to have a mostly direct translation of commands that won't work out of the box, so that those additional repositories can be added in the proper templates without complication. solution is to have those templates warn the user when they need to be updated, with the same expected warnings of templates to be updated, and same GUI UX warning them that they should restart their qubes consequently.

The solution presented here is creating a new curl-proxy command to be used by users into replacing wget/curl calls. Simple fix to a really big UX problem. See the following signal-desktop/session-desktop examples (not perfect here, as you will see withsignal-desktop, the wrapper should parse what is given and maybe clean it prior of redirecting input. `-O-` was manually removed: the user would just don't know.)

Thanks again for your work in UX!!!!

in debian-11 template:
`sudo apt install curl`
# curl-proxy
```
#!/bin/env bash
curl --proxy http://127.0.0.1:8082/ --tlsv1.2 --proto =https --max-time 180 "$@"
```

# wget-proxy
```
#!/bin/env bash
https_proxy=http://127.0.0.1:8082/ http_proxy=http://127.0.0.1:8082/ wget --secure-protocol=TLSv1_2 --timeout=180 "$@"
```

#### Session-desktop documentation (https://deb.oxen.io/) drop in replacement:

```
sudo curl -so /etc/apt/trusted.gpg.d/oxen.gpg https://deb.oxen.io/pub.gpg
echo "deb https://deb.oxen.io $(lsb_release -sc) main" | sudo tee /etc/apt/sources.list.d/oxen.list
sudo apt update
sudo apt install session-desktop
```

Here the user would have to replace:
`sudo curl -so /etc/apt/trusted.gpg.d/oxen.gpg https://deb.oxen.io/pub.gpg`
to
`sudo curl-proxy -so /etc/apt/trusted.gpg.d/oxen.gpg https://deb.oxen.io/pub.gpg`

#### Signal-desktop instructions (https://signal.org/download/linux/) drop-in replacement:
Upstream instructions:

```
wget -O- https://updates.signal.org/desktop/apt/keys.asc | gpg --dearmor > signal-desktop-keyring.gpg
cat signal-desktop-keyring.gpg | sudo tee -a /usr/share/keyrings/signal-desktop-keyring.gpg > /dev/null
echo 'deb [arch=amd64 signed-by=/usr/share/keyrings/signal-desktop-keyring.gpg] https://updates.signal.org/desktop/apt xenial main' |\
  sudo tee -a /etc/apt/sources.list.d/signal-xenial.list
sudo apt update && sudo apt install signal-desktop
```

Here the following line needs to be changed:
`wget -O- https://updates.signal.org/desktop/apt/keys.asc | gpg --dearmor > signal-desktop-keyring.gpg`
becomes
`wget-proxy -O- https://updates.signal.org/desktop/apt/keys.asc | gpg --dearmor > signal-desktop-keyring.gpg`

### The value to a user, and who that user might be
Be able to follow a single Qubes documentation explaining the conversion from instructions read online requiring to download gpg keys, why it won't work and what to replace by what.
Be able to be autonomous into installing software.

Edit: added wget-proxy based on additional experimentations and discussions that happened under https://forum.qubes-os.org/t/installing-software-in-qubes-all-methods/9991/10

**Comment by tlaurion** [comment]
@marmarek @ninavizz ? Others?

**Comment by DemiMarie** [comment]
@tlaurion you need to add double quotes around the `$@`, FYI

**Comment by tlaurion** [comment]
@DemiMarie Thanks for the pointer and corrected in proposition.

But idea is more about how to be able to propose a drop in replacement for all users trying to follow different guides out there, and always falling short. Any ideas welcome to make this better for everyone.

**Comment by andrewdavidwong** [comment]
Not sure if we should be encouraging curl-pipe-anything, but I agree that users are struggling with how to install software that's not available in the default repos.

**Comment by logoerthiner1** [comment]
I wonder what is the recommended way in Qubes OS if I want to install a python package from pip.

**Comment by DemiMarie** [comment]
> I wonder what is the recommended way in Qubes OS if I want to install a python package from pip.

Do it in an AppVM

**Comment by deeplow** [comment]
> @deeplow has been working on community documentation that covers how to install such software.

Just dropping here a link to that https://forum.qubes-os.org/t/9991/1 in case someone wonders where it is.

**Comment by tlaurion** [comment]
@deeplow @DemiMarie @andrewdavidwong : posted comment here that links here:
https://forum.qubes-os.org/t/installing-software-in-qubes-all-methods/9991/10

**Comment by deeplow** [comment]
I really like this idea. @tlaurion, why not a simple alias in `.bashrc`?

**Comment by deeplow** [comment]
Also, a potential issue with the proposed commands is specifying the TLS version. The developers will need to keep tabs on the deprecation state of this version and update it.

Maybe there's a way for it to automatically negotiate the strongest encryption? (haven't checked)

**Comment by tlaurion** [comment]
> I really like this idea. @tlaurion, why not a simple alias in .bashrc?

> Also, a potential issue with the proposed commands is specifying the TLS version. The developers will need to keep tabs on the deprecation state of this version and update it.

@deeplow Qubes intentionally hides network access of Templates behind a proxy. The idea of creating wrapper scripts is for the user to intentionally and explicitly type that replacement command. 

Putting an alias would defeat the protection, permitting whatever being automatized by a script to do damage without being witnessed by a user. In such case where obscure and unaudited scriplet is intended to be used by a user, which runs said sniplet which fails, maybe additional documentation could be given to define such aliases at console, so that the aliases would be valid only for that launched terminal? Otherwise, defining aliases would permit anything using curl/wget to download whatever needed by anything relying on those tools automatically, defeating the purpose of the update proxy used by templates internally, wrapped solely in OS updater as of now (dnf/apt etc).

As for TLS1.2 being pinned here, the idea was borrowed from Whonix for the curl command I have seen once hidden in some documentation. The idea here is to force usage of known good protocol in a whitelisting approach. 

wget permits to list all permitted (we could extend the list) where no blacklisting approach exists. Agreed it would add a bit if maintainership for the script in the future, but the whole idea is to block less secure protocol, where exploits permitted in the past to renegociate less secure protocols inside of an established session, which is what we try to avoid here, mainly mitm attacks. By forcing https and denying http, pinning at least known good tlsv1.2, its a good start but could of course be improved by enforcing better defaults according to wget/curl options.

First I tried to have curl replace wget calls to have a single wrapper to rule them all. But then, the wrapper would need to be more intelligent then its own good (-O- in Signal installation guide exemple will not fly when used with curl).

Doing a real quick research on installation guides available combined with past customers questions asked on software desired to be installed, it seems that having curl and wget wrapper would fit the bill and bypass gpg key download problems.

@DemiMarie suggested to deploy a Qubes package containing keys of mostly used distro signing keys of software repositories. I like the idea, but there will always be corner cases from users attempting to install xyz or having the need to download zxy, which will go back to the need of having such tool.

I think the wrappers should probzbly log URL accessed though. Or the proxy logs being brought to the attention of the user when the updates are not triggered by Qubes?
Not sure how to make this convenient and not less secure at the same time. Once wget-wrapper/curl-wrapper is there, its there. Seems like the best option is the logging on the proxy side, while having some oversight from Qubes internals to bring attention to what happens there, so that UX convenience is met.

All of this may be simply documented as well though. A user installing custom channels might be enticed by documentation to learn to dig on how to inspect the proxy logs, and also learn better opsec at the same time.

@andrewdavidwong thoughts?

**Comment by andrewdavidwong** [comment]
> All of this may be simply documented as well though. A user installing custom channels might be enticed by documentation to learn to dig on how to inspect the proxy logs, and also learn better opsec at the same time.
> 
> @andrewdavidwong thoughts?

I'm generally in favor of separating [core and external documentation](https://www.qubes-os.org/doc/documentation-style-guide/#core-vs-external-documentation). "How to install software in templates" would be core, while "how to install software in Linux" would be external. For example, all the stuff about curl, wget, and GPG would be external, while the parts about qvm-copying from a disposable to a template would be core. If the desire is to gave a single unified guide, it should be external.

**Comment by tlaurion** [comment]
For those not following forum thread, Software installation instructions have been updated by @andrewdavidwong [here](https://forum.qubes-os.org/t/curl-proxy-wget-proxy-scripts-in-templates-so-users-can-add-gpg-distro-keys-linked-to-added-external-repositories/10935/32?u=insurgo)

**Comment by tlaurion** [comment]
The idea was vastly rejected in favor of a [MOTD educating the user.](https://forum.qubes-os.org/t/curl-proxy-wget-proxy-scripts-in-templates-so-users-can-add-gpg-distro-keys-linked-to-added-external-repositories/10935/76?u=insurgo)

The discussion evolved into going back and understanding why [tinyproxy lost its filtering capabilities](https://forum.qubes-os.org/t/why-does-the-updateproxy-not-have-a-whitelist/11087/2?u=insurgo).

**Comment by ubuntupunk** [comment]
bump, I just want to install windsurf in a persistent state i.e template qube without doing something nuts like copying it over from an AppVM.

**Comment by tlaurion** [comment]
> bump, I just want to install windsurf in a persistent state i.e template qube without doing something nuts like copying it over from an AppVM.

@ubuntupunk this issue is closed as "non planned". This was a long discussed topic over the forum, where final points were to create a motd warning user, and guiding user into punching holes of some sort to add additional repos and download additional gpg public keys into a template. Nothing was good enough. 


Recently, @marmarek and @marmarta looked a bit into extrepo and extrepo-data-offline debian packages, which aims at providing helpers to deploy additional repositories and associated public gpg key.

I'm not aware that an issue is opened to track that work, now could be a good time to create such issue.

I do not know if package xyz is supported by extrepo project, that would be left to users to check if xyz supports debian packaging and repos, and willing contributors to contribute under extrepo git project to open issue, or better, propose merge requests to add additional software repo deployment there.

Notes : not all open source projects prepare debian repositories. Some insist into deploying flatpak, appimages or snap statically linked binaries all contained in a big fat executable. This is another subject which won't be resolved by extrepo which aims only at helping adding known repositories and their associated public keys. I would advise using appimages and flatpak as last resorts, where snap are at least kinda properly supported by qubesos with update hooks permitting to keep templates and their snap up to date as similar to software repositories and warn when there is an update available, permitting you to update your repo before starting your work session if it's been a while you haven't used your computer, as part of a good opsec hygiene. 

One step woukd be to look if your xyz desired packages have debian repository and if so, open at least an issue under extrepo issue tracker pointing to the installation instructions.

Hope this helps.

**Comment by ubuntupunk** [comment]
I browsed this thread, and its long. The extrepo suggestion is right at the end, and it doesn't come up when googling the issue. Is there a qubes wiki where the correct procedure for adding additional repos & gpg keys is outlined for ordinary users, in simple terms? We have to create some level of trust and the trust is all about the gpg keys IMHO, beyond that, nobody can guarantee anything further.
